### PR TITLE
fix type-stability bugs in Ryu code

### LINKED ISSUE #52749: printf MethodError with "append_nine_digits"
This seems like a bug:
```jl
julia> using Printf

julia> @printf("%.60g", 1.38e-23)
1.3800000000000000600105824657340787992976609667826426243954e-23
julia> @printf("%.160g", 1.38e-23)
ERROR: MethodError: no method matching append_nine_digits(::Int64, ::Vector{UInt8}, ::Int64)

Closest candidates are:
  append_nine_digits(::Unsigned, ::Any, ::Int64)
   @ Base intfuncs.jl:815

Stacktrace:
 [1] writeexp(buf::Vector{…}, pos::Int64, v::Float64, precision::Int64, plus::Bool, space::Bool, hash::Bool, expchar::UInt8, decchar::UInt8, trimtrailingzeros::Bool)
   @ Base.Ryu ./ryu/exp.jl:113
 [2] writeexp(buf::Vector{…}, pos::Int64, v::Float64, precision::Int64, plus::Bool, space::Bool, hash::Bool, expchar::UInt8, decchar::UInt8, trimtrailingzeros::Bool)
   @ Base.Ryu ./ryu/exp.jl:4 [inlined]
 [3] fmt
   @ ~/Documents/Code/julia/usr/share/julia/stdlib/v1.11/Printf/src/Printf.jl:531 [inlined]
 [4] fmt
   @ ~/Documents/Code/julia/usr/share/julia/stdlib/v1.11/Printf/src/Printf.jl:313 [inlined]
 [5] format
   @ ~/Documents/Code/julia/usr/share/julia/stdlib/v1.11/Printf/src/Printf.jl:834 [inlined]
 [6] format(io::Base.TTY, f::Printf.Format{Base.CodeUnits{UInt8, String}, Tuple{Printf.Spec{Val{'g'}}}}, args::Float64)
   @ Printf ~/Documents/Code/julia/usr/share/julia/stdlib/v1.11/Printf/src/Printf.jl:935
```

It looks like the `Unsigned` type qualifier was added by @vtjnash to `append_nine_digits` in #51273, so it's a pretty recent problem.

### PR DESCRIPTION
**Pull request by stevengj**
Fixes #52749.

### CODE REVIEW

File: base/ryu/exp.jl [hunk 1]
```diff
@@ -197,7 +197,7 @@ function writeexp(buf, pos, v::T,
                 roundUp = 1
                 continue
             else
-                if roundUp == 2 && UInt8(c) % 2 == 0
+                if roundUp == 2 && iseven(UInt8(c))
```
Comment (stevengj) [inline base/ryu/exp.jl:200]:
```suggestion
                if roundUp == 2 && iseven(c)
```

File: base/ryu/exp.jl [hunk 2]
```diff
@@ -224,7 +224,7 @@ function writeexp(buf, pos, v::T,
         pos += 1
     end
     if e >= 100
-        c = e % 10
+        c = UInt8(e % UInt8(10))
```
Comment (stevengj) [inline base/ryu/exp.jl:227]:
```suggestion
        c = UInt8(e % 10)
```

File: base/ryu/exp.jl [hunk 3]
```diff
@@ -224,7 +224,7 @@ function writeexp(buf, pos, v::T,
         pos += 1
     end
     if e >= 100
-        c = e % 10
+        c = UInt8(e % 10)
```
Comment (stevengj) [inline base/ryu/exp.jl:227]:
```suggestion
        c = (e % 10) % UInt8
```

**Review by vtjnash — APPROVED** [review]

**Comment by stevengj** [comment]
CI failures are unrelated; should be good to merge.

**Pull request merged**